mongo/src/main/java/com/redhat/lightblue/mongo/metadata/BSONParser.java

```diff
@@ -51,140 +51,86 @@
 import com.redhat.lightblue.util.JsonUtils;
 
 @SuppressWarnings({"rawtypes", "unchecked"})
-public class BSONParser extends MetadataParser<BSONObject> {
+public class BSONParser extends MetadataParser<Object> {
     private static final Logger LOGGER = LoggerFactory.getLogger(BSONParser.class);
 
     public static final String DELIMITER_ID = "|";
 
-    public BSONParser(Extensions<BSONObject> ex,
+    public BSONParser(Extensions ex,
                       TypeResolver resolver) {
         super(ex, resolver);
     }
 
     @Override
-    public String getStringProperty(BSONObject object, String name) {
-        Object x = object.get(name);
-        if (x != null) {
-            if (x instanceof String) {
-                return (String) x;
-            } else {
-                throw Error.get(MetadataConstants.ERR_ILL_FORMED_METADATA, name);
-            }
-        } else {
-            return null;
-        }
-    }
-
-    @Override
-    public BSONObject getObjectProperty(BSONObject object, String name) {
-        Object x = object.get(name);
-        if (x != null) {
-            if (x instanceof BSONObject) {
-                return (BSONObject) x;
-            } else {
-                throw Error.get(MetadataConstants.ERR_ILL_FORMED_METADATA, name);
-            }
-        } else {
-            return null;
-        }
-    }
-
-    @Override
-    public Object getValueProperty(BSONObject object, String name) {
-        Object x = object.get(name);
-        if (x != null) {
-            if (x instanceof Number
-                    || x instanceof String
-                    || x instanceof Date
-                    || x instanceof Boolean) {
-                return x;
-            } else {
-                throw Error.get(MetadataConstants.ERR_ILL_FORMED_METADATA, name);
-            }
-        } else {
-            return null;
-        }
-    }
-
-    @Override
-    public List<String> getStringList(BSONObject object, String name) {
-        Object x = object.get(name);
-        if (x != null) {
-            if (x instanceof List) {
-                ArrayList<String> ret = new ArrayList<>();
-                for (Object o : (List) x) {
-                    ret.add(o.toString());
-                }
-                return ret;
-            } else {
-                throw Error.get(MetadataConstants.ERR_ILL_FORMED_METADATA, name);
-            }
-        } else {
-            return null;
-        }
+    public Object newMap() {
+        return new BasicDBObject();
```

Comment:
Yes, but the metadataparser was incorrectly parameterized. It takes only
one parameter <T>, and assumes the metadata is represented using objects
all deriving from T. This is correct for JSON, where everyrhing is a
JsonNode, but it is not correct for BSON, because maps are DBObject, values
are Java values and lists are java Lists. So, Object is the only common
denominator for BSON.

The parser should have been parameterized using <L, M, V>, different types
for list, map, and value.

On Fri, Apr 1, 2016 at 9:27 AM, Dennis Crissman <notifications@github.com>
wrote:

> In mongo/src/main/java/com/redhat/lightblue/mongo/metadata/BSONParser.java
> <https://github.com/lightblue-platform/lightblue-mongo/pull/232#discussion_r58221972>
> :
>
> > -        Object x = object.get(name);
> > -        if (x != null) {
> > -            if (x instanceof List) {
> > -                ArrayList<String> ret = new ArrayList<>();
> > -                for (Object o : (List) x) {
> > -                    ret.add(o.toString());
> > -                }
> > -                return ret;
> > -            } else {
> > -                throw Error.get(MetadataConstants.ERR_ILL_FORMED_METADATA, name);
> > -            }
> > -        } else {
> > -            return null;
> > -        }
> > +    public Object newMap() {
> > +        return new BasicDBObject();
>
> I won't block on this, but as a rule of thumb, I try to return Objects at
> the highest level possible and pass in parameters at the lowest level as
> possible. In this case, if there were a use case to use BSONParser
> directly, then casting wouldn't be required.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly or view it on GitHub
> <https://github.com/lightblue-platform/lightblue-mongo/pull/232/files/b6bba7e856c5d1d000a696ddd1d634f844bf77c1#r58221972>
>